### PR TITLE
optimize advancedsubtensor

### LINKED ISSUE #3166: Convert AdvancedSubtensor to GpuAdvancedSubtensor1 on the GPU
@lamblin you told me you know how to make this. Can you describe it here so that someone else have a change to make it?

### PR DESCRIPTION
**Pull request by yingzha**
Fixes #3166.
@abergeron, I didn't push the test function for this optimization since when I tested it locally, it said optimization failed and I'm not sure which part is going wrong.
I did not remove the annotation to make it a bit clearer. Could you take a look on the function and notify me where the problem is?
Thanks!
- [x] Optimization function
- [ ]  Test function for it

### CODE REVIEW

File: theano/tensor/opt.py [hunk 1]
```diff
@@ -3231,6 +3233,124 @@ def local_adv_sub1_adv_inc_sub1(node):
...
+    #get complete dimshuffle information
+    dimshuffle_idx = pre_dim_idx + va_dim_idx + post_dim_idx
+    x = x.dimshuffle(*dimshuffle_idx)
+    end_va = end - len(va_s_idx)
+    start_va = start + len(nx_idx)
+
+    #step 3: partial flattening
+    shape = x.shape[: start_va] +
+            (T.prod(x.shape[start_va: end_va]),) +
+            x.shape[end_va:]
```
Comment (lamblin) [inline theano/tensor/opt.py]: You need parentheses around this expression if you want to use multiple lines

File: theano/tensor/opt.py [hunk 2]
```diff
@@ -3231,6 +3233,124 @@ def local_adv_sub1_adv_inc_sub1(node):
...
+    #step 6: advanced slicing
+    out_flat = input_flat.take(new_idx.flatten())
+
+    #step 7: reshape into right shape
+    o = out_flat.reshape(x.shape[:start_va] +
+            new_idx.shape + x.shape[end_va:], ndim=dims)
+
+    idx_ = [slice(None) if i == 'x' else idx[i] for i in pre_dim_idx] +
+           [slice(None)] * (end_va - start) + [idx[i] for i in s_idx] +
+           [slice(None) if i == 'x'else idx[i] for i in post_dim_idx]
```
Comment (lamblin) [inline theano/tensor/opt.py]: Same here.

**Comment by yingzha** [comment]
@abergeron, here is the test. Before going to the optimization part, the original slice type are converted MakeSlice type then when I tried to build regular subtensor, it failed.

def test_local_adv_sub():
     shp = (2, 3, 4, 5, 6)
     xval = numpy.arange(numpy.prod(shp), dtype=theano.config.floatX).reshape(shp)
     idx1, idx2 = tensor.ivectors('idx1', 'idx2')
     idxs = [slice(None), idx1, None, slice(None), slice(0, 2, 1),idx2, None]
     x = theano.shared(xval, name='x')
     mode = compile.get_default_mode().including("local_adv_sub")
     expr = x[idxs]
     f = theano.function([idx1, idx2], expr, mode=mode)
     topo = f.maker.fgraph.toposort()

**Comment by yingzha** [comment]
@abergeron, here is the test. Before going to the optimization part, the original slice type are converted MakeSlice type then when I tried to build regular subtensor, it failed.

`def test_local_adv_sub():
shp = (2, 3, 4, 5, 6)
xval = numpy.arange(numpy.prod(shp), dtype=theano.config.floatX).reshape(shp)
idx1, idx2 = tensor.ivectors('idx1', 'idx2')
idxs = [slice(None), idx1, None, slice(None), slice(0, 2, 1),idx2, None]
x = theano.shared(xval, name='x')
mode = compile.get_default_mode().including("local_adv_sub")
expr = x[idxs]
f = theano.function([idx1, idx2], expr, mode=mode)
topo = f.maker.fgraph.toposort()`

**Comment by yingzha** [comment]
@abergeron, here is the test. Before going to the optimization part, the original slice type are converted MakeSlice type then when I tried to build regular subtensor, it failed.

`def test_local_adv_sub():`
`shp = (2, 3, 4, 5, 6)`
`xval = numpy.arange(numpy.prod(shp), dtype=theano.config.floatX).reshape(shp)`
`idx1, idx2 = tensor.ivectors('idx1', 'idx2')`
`idxs = [slice(None), idx1, None, slice(None), slice(0, 2, 1),idx2, None]`
`x = theano.shared(xval, name='x')`
`mode = compile.get_default_mode().including("local_adv_sub")`
`expr = x[idxs]`
`f = theano.function([idx1, idx2], expr, mode=mode)`
`topo = f.maker.fgraph.toposort()`

**Pull request closed without merging**